src/constraints.jl

```diff
@@ -628,15 +545,66 @@ function _error_if_not_concrete_type(t)
     end
     return
 end
+# `isconcretetype(Vector{Integer})` is `true`
+function _error_if_not_concrete_type(t::Type{Vector{ElT}}) where ElT
+    _error_if_not_concrete_type(ElT)
+end
+
+"""
+    num_constraints(model::Model, function_type, set_type)::Int64
+
+Return the number of constraints currently in the model where the function
+has type `function_type` and the set has type `set_type`.
+
+See also [`list_of_constraint_types`](@ref) and [`list_of_constraints`](@ref).
+
+# Example
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x >= 0, Bin);
+
+julia> @variable(model, y);
+
+julia> @constraint(model, y in MOI.GreaterThan(1.0));
+
+julia> @constraint(model, y <= 1.0);
+
+julia> @constraint(model, 2x <= 1);
+
+julia> num_constraints(model, VariableRef, MOI.GreaterThan{Float64})
+2
+
+julia> all_constraints(model, VariableRef, MOI.ZeroOne)
+1
+
+julia> all_constraints(model, AffExpr, MOI.LessThan{Float64})
+2
+```
+"""
+function num_constraints(
+    model::Model,
+    function_type::Type{<:Union{AbstractJuMPScalar,
+                                Vector{<:AbstractJuMPScalar}}},
+    set_type::Type{<:MOI.AbstractSet})::Int64
+    _error_if_not_concrete_type(function_type)
+    _error_if_not_concrete_type(set_type)
+    # TODO: Support JuMP's set helpers like SecondOrderCone().
+    f_type = moi_function_type(function_type)
+    return MOI.get(model, MOI.NumberOfConstraints{f_type, set_type}())
+end
+
 
 """
-    all_constraints(model::Model, function_type, set_type)::Vector{VariableRef}
+    all_constraints(model::Model, function_type, set_type)::Vector{
+        ConstraintRef{Model,
+            MOI.ConstraintIndex{moi_function_type(function_type), set_type}}}
```

Comment:
It's pretty ugly to list this return type. Not sure what the right way to handle this is.

Reply:
I wasn't sure either, @odow any idea ?

Reply:
I have simplified it. Since the shape is different from vector to scalar, it is too complex to put in a signature